### ISSUE TITLE
OpenMP 5 back-end broken for clang 12

### ISSUE THREAD

**Issue by j-stephan** [alpaka-group/alpaka#1381]
Compiling the alpaka tests with clang 12 and an activated OpenMP 5 back-end leads to compilation errors:

```
In file included from /home/stepha27/workspace/caravan/alpaka/test/integ/sharedMem/src/sharedMem.cpp:10:
In file included from /home/stepha27/workspace/caravan/alpaka/include/alpaka/alpaka.hpp:112:
/home/stepha27/workspace/caravan/alpaka/include/alpaka/kernel/TaskKernelOmp5.hpp:191:29: error: Type 'std::tuple<int *>' is not trivially copyable and not guaranteed to be mapped correctly [-Werror,-Wopenmp-mapping]
                            argsD);
                            ^~~~~
```

The culprit is this member variable of `TaskKernelOmp5`:

https://github.com/alpaka-group/alpaka/blob/c938c8045bb9592fc1ee6a05a440d3cfc279c2d3/include/alpaka/kernel/TaskKernelOmp5.hpp#L206

I didn't test this with clang 11 yet. In clang 10 this works because our CI doesn't complain.

**Comment by jkelling** [comment]
The issue is, that `-Wno-openmp-mapping` is not added in cmake, but explicitly set via `CMAKE_CXX_FLAGS` in the ci environment. When clang12 was added to the CI this bit was probably omitted.

This issue should be about, whether this backend should be properly integrated in alpaka's cmake config. At the moment `CMAKE_CXX_FLAGS` must be set manually to select the target and set the target GPU arch if the target is GPU. I do not know how to best represent this in cmake, and whether we should do this at this point.

**Comment by jkelling** [comment]
The standard does not demand `std::tuple` be trivially copyable if the component types are. The gcc implementation has defaulted copy and move constructors but custom assignments, which makes it not fulfill `is_trivially_copyable`. It does not allocate anything on the heap, so it is effectively `memcpy`-able but not formally. I.e.: it works when turning off the warning, but it would be cleaner to replace `std::tuple` with a compatible tuple which is guaranteed to be trivially copyable if all component types are.

However: earlier the check behind this warning also required the type to be trivially constructible, too. I did no check if the changed that in 12, back with 10, they argued against changing that bit. Thus the warning may not go away even with a clean tuple.

**Comment by bernhardmgruber** [comment]
The behavior in clang was introduced here and it is a pure diagnostic change: https://reviews.llvm.org/D71134
I am for disabling the warning.

**Comment by sbastrakov** [comment]
Note on a related topic: we mention that user's `Kernel` types are required to be trivially copyable in two places in the docs: [one](https://github.com/alpaka-group/alpaka/blob/develop/docs/source/basic/library.rst), [two](https://github.com/alpaka-group/alpaka/blob/develop/docs/source/advanced/rationale.rst). But such wording implies we should be checking it in alpaka, like we used to do. So we should either add a check or relax the wording (e.g. to `should`, not `has to`) in the docs.

**Comment by j-stephan** [comment]
I already encountered this with SYCL. Easy solution: https://github.com/codeplaysoftware/computecpp-sdk/blob/master/include/stl-tuple/STLTuple.hpp

This is from a SYCL SDK, but doesn't introduce any SYCL dependency. It is just a dumber tuple type that is suitable for device code.

**Comment by bernhardmgruber** [comment]
I don't think the right solution is to pick a different tuple implementation. `std::tuple` should be fixed by making it trivially copyable if all contained types are trivially copyable. And we should disable the diagnostic as actually using `std::tuple` does not harm.

**Comment by j-stephan** [comment]
> `std::tuple` should be fixed by making it trivially copyable if all contained types are trivially copyable.

That is hardly something we can do.

> And we should disable the diagnostic as actually using std::tuple does not harm.

If this is possible without interfering with user code, I'm all for it.

**Comment by sbastrakov** [comment]
Yes, we can't directly change `std::tuple` implementations, only complain about it and hope for a change. However, we can easily make out own `IsTriviallyCopyable` trait that is by default same as its `std::` counterpart, but for an `std::tuple` of trivially copyable types returns `true`.

**Comment by jkelling** [comment]
> The behavior in clang was introduced here and it is a pure diagnostic change: https://reviews.llvm.org/D71134
> I am for disabling the warning.

This commit actually removes the `isTrivial` check and reduces it to `isTriviallyCopyable`, this speaks for keeping the warning enabled, as the check is sensible albeit strict.

**Comment by jkelling** [comment]
> Yes, we can't directly change `std::tuple` implementations, only complain about it and hope for a change. However, we can easily make out own `IsTriviallyCopyable` trait that is by default same as its `std::` counterpart, but for an `std::tuple` of trivially copyable types returns `true`.

This would not address this issue, as the check underlying this warning is done by clang using its own `isTriviallyCopyableType`.

**Comment by sbastrakov** [comment]
Yes, it won't solve this warning. It would be helpful if we add our old check back.

**Comment by j-stephan** [comment]
We could first check the individual arguments for `IsTriviallyCopyable` and then silence the warning around the places where the tuple is used. Would that make sense?

**Comment by jkelling** [comment]
> We could first check the individual arguments for `IsTriviallyCopyable` and then silence the warning around the places where the tuple is used. Would that make sense?

Yes it would. Using that simpler tuple class would be cleaner though. Is this included in the alpaka source tree by https://github.com/alpaka-group/alpaka/pull/789 ?

Passing the arguments as a parameter pack into the target region does not work because then the compiler sees any naked pointers in there and tries to map them. If there is not mapping for an address to target memory (e.g. because it is already a device address) then Clang 14(main) decides to not copy the pointer and replace it by `0` instead. The `is_device_ptr` clause exists to tell the compiler, that something is already a device address, but as we only have a parameter pack, we cannot use this. The parameters need to be packed into some structure to hide any device pointers form the compiler.

**Comment by j-stephan** [comment]
> Is this included in the alpaka source tree by #789 ?

No, we require this as a third-party dependency if SYCL is activated.

**Comment by jkelling** [comment]
I would prefer the simple-tuple solution over checking the argument types and silencing the warning. The main argument for this is, that the OpenMP5 backend is currently used less for production use than to chase compiler bugs. Ignoring a compiler warning will always raise red flags and distract from the actual issues. It would be better to have something where the compiler does not complain before it crashes.

Could we include that [`STLTuple`](https://github.com/codeplaysoftware/computecpp-sdk/blob/master/include/stl-tuple/STLTuple.hpp) in alpaka somewhere under `include/`, so we do not need to have an external dependency? [MPL](https://www.mozilla.org/en-US/MPL/2.0/FAQ/) does appear to be compatible with Apache this way:
```
Q13: May I combine MPL-licensed code and BSD-licensed code in the same executable program? What about Apache?

Yes to both. Mozilla currently does this with BSD-licensed code. For example, libvpx, which is used in Firefox to decode WebM video, is under a BSD license.
```
